rplugin/python3/deoplete/sources/beancount.py

```diff
@@ -0,0 +1,75 @@
+import re
+
+from deoplete.source.base import Base
+
+try:
+    from beancount.loader import load_file
+    from beancount.core.data import Open, Transaction
+    HAS_BEANCOUNT = True
+except ImportError:
+    HAS_BEANCOUNT = False
+
+DIRECTIVES = ['open', 'close', 'commodity', 'txn', 'balance', 'pad', 'note',
+              'document', 'price', 'event', 'query', 'custom']
+
+
+class Source(Base):
+    def __init__(self, vim):
+        super().__init__(vim)
+        self.vim = vim
+
+        self.name = 'beancount'
+        self.mark = '[bc]'
+        self.filetypes = ['beancount']
+        self.min_pattern_length = 0
+
+    def on_init(self, context):
+        if not HAS_BEANCOUNT:
+            self.error('Importing beancount failed.')
+
+    def on_event(self, context):
+        self.__make_cache(context)
+
+    def get_complete_position(self, context):
+        m = re.search(r'\S*$', context['input'])
+        return m.start() if m else -1
+
+    def gather_candidates(self, context):
+        if re.match(r'^\d{4}[/-]\d\d[/-]\d\d \w*$', context['input']):
+            return DIRECTIVES
+        if not context['complete_str']:
+            return []
+        first = context['complete_str'][0]
+        if first == '#':
+            return ['#' + w for w in self._tags]
+        elif first == '^':
+            return ['^' + w for w in self._links]
+        elif first == '"':
+            return ['"{}"'.format(w) for w in self._payees]
+        return self._accounts
```

Comment:
It would be really cool to add partial account matching, so that you can work your way up the tree in steps instead of having to match the entire account name all at once (eg "as<tab>:bro<tab>:ho<tab>" for "Assets:Brokerage:Hooli").

Unless that's what this actually does... I haven't tested it yet, but I will do it as soon as I update my container to include neovim.